tests/Unit/DataStructures/Tensor/Expressions/Test_Contract.cpp

```diff
@@ -0,0 +1,744 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/Contract.hpp"
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/Requires.hpp"
+
+template <typename T,
+          Requires<std::is_same_v<typename T::type, DataVector>> = nullptr>
+void assign_unique_datavector_tensor_values(T& tensor) {
+  double value = 0.0;
+  for (auto index_it = tensor.begin(); index_it != tensor.end(); index_it++) {
+    for (auto vector_it = index_it->begin(); vector_it != index_it->end();
+         vector_it++) {
+      *vector_it = value;
+      value += 1.0;
+    }
+  }
+}
+
+SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.Expression.Contract",
+                  "[DataStructures][Unit]") {
+  // Each test case here is run twice: once with a Tensor that store doubles
+  // and a second time with a Tensor that store DataVectors
+  //
+  // Contract rank 2 (upper, lower) tensor to rank 0 tensor
+  // Use explicit type (vs auto) so the compiler checks the return type of
+  // `evaluate`
+  Tensor<double, Symmetry<2, 1>,
+         index_list<SpatialIndex<3, UpLo::Up, Frame::Inertial>,
+                    SpatialIndex<3, UpLo::Lo, Frame::Inertial>>>
+      Rul{};
+  std::iota(Rul.begin(), Rul.end(), 0.0);
+
+  const Tensor<double> RIi_contracted =
+      TensorExpressions::evaluate(Rul(ti_I, ti_i));
+
+  double expected_RIi_sum = 0.0;
+  for (size_t i = 0; i < 3; i++) {
+    expected_RIi_sum += Rul.get(i, i);
+  }
+  CHECK(RIi_contracted.get() == expected_RIi_sum);
+
+  Tensor<DataVector, Symmetry<2, 1>,
+         index_list<SpatialIndex<3, UpLo::Up, Frame::Inertial>,
+                    SpatialIndex<3, UpLo::Lo, Frame::Inertial>>>
+      Sul(4_st);
+  assign_unique_datavector_tensor_values(Sul);
```

Comment:
When `Tensor`s with `DataVector`s are initialized in the way above this line, and then `std::iota` is used to assign the elements (like [this](https://github.com/sxs-collaboration/spectre/blob/5d37e9c2387997e0385bfe502f1fa20b0708235e/tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp#L48)), each individual `DataVector` in the `Tensor`  generated will have the same value for each of its elements (e.g. the `DataVector` s are `{0, 0, 0}`, `{1, 1, 1}`, `{2, 2, 2}` ... ). If I assigned the elements this way, I figured that checking that the contracted `Tensor` has the correct element-wise sums is not very rigorous because the element-wise sum is the same for each `DataVector` index.

To address this, I wrote this `assign_unique_datavector_tensor_values` function that will assign unique values to each element of the `Tensor`, across all `DataVector`s. It is similar to if all the `DataVector`s were contiguous and `std::iota`  was applied across them. I'm not sure if doing something like this is necessary or beneficial. If it is, let me know if this function could be improved upon or helpful elsewhere in the code to be used more generally and not just within this test.